tests/explainers/test_tree.py

```diff
@@ -1197,6 +1197,30 @@ def test_xgboost_regression(self):
         expected_diff = np.abs(explanation.values.sum(1) + explanation.base_values - predicted).max()
         assert expected_diff < 1e-4, "SHAP values don't sum to model output!"
 
+    def test_xgboost_dmatrix_propagation(self):
+        xgboost = pytest.importorskip("xgboost")
+
+        def sigmoid(x):
+            return 1 / (1 + np.exp(-x))
+
+        X, y = shap.datasets.adult()
+
+        # Randomly set some features of X to NaN to test propagation of missing param
+        X_nan = X.copy()
+        X_nan.loc[
+            X_nan.sample(frac=0.3, random_state=42).index,
+            X_nan.columns.to_series().sample(frac=0.5, random_state=42),
+        ] = 1e-8
+
+        clf = xgboost.XGBClassifier(missing=1e-8, random_state=42)
+        clf.fit(X_nan, y)
+        pred_probs = clf.predict_proba(X_nan)[:, 1]
+        explainer = shap.TreeExplainer(clf)
+        shap_values = explainer.shap_values(X_nan)
+        # check that SHAP values sum to model output
+        expected_diff = np.max((pred_probs - sigmoid(explainer.expected_value + shap_values.sum(axis=1))).abs())
+        assert expected_diff < 1e-5, "SHAP values don't sum to model output!"
```

Comment:
I'm thinking if it's really necessary to do the comparison in probability space? Would it be easier to compare in margin space, then we don't even need the sigmoid? (You can refer to the other xgboost tests for an example)

Reply:
It makes it easier to do it in the margin space so I have amended the test to do so. The reason it was in the probaiblity space is because that is how I uncovered the bug when using `predict_proba` from the sklearn-like `xgboost.sklearn.XGBClassifier` implementation and comparing it to the lower level `predict` method on the `xgboost.core.Booster` which is what shap was using. The bug still exists either way since it was related to the DMatrix implementation in the `Tree` class of shap.